lib/srv/discovery/discovery.go

```diff
@@ -795,13 +798,108 @@ func (s *Server) handleEC2Instances(instances *server.EC2Instances) error {
 	// instances.Rotation is true whenever the instances received need
 	// to be rotated, we don't want to filter out existing OpenSSH nodes as
 	// they all need to have the command run on them
-	if !instances.Rotation {
+	//
+	// Integration/EICE Nodes don't have heartbeat.
+	// Those Nodes must not be filtered, so that we can extend their expiration and sync labels.
+	if !instances.Rotation && instances.Integration == "" {
 		s.filterExistingEC2Nodes(instances)
 	}
 	if len(instances.Instances) == 0 {
 		return trace.NotFound("all fetched nodes already enrolled")
 	}
 
+	switch {
+	case instances.Integration != "":
+		s.heartbeatEICEInstance(instances)
+	default:
+		if err := s.handleEC2RemoteInstallation(instances); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if err := s.emitUsageEvents(instances.MakeEvents()); err != nil {
+		s.Log.WithError(err).Debug("Error emitting usage event.")
+	}
+
+	return nil
+}
+
+// heartbeatEICEInstance heartbeats the list of EC2 instances as Teleport (EICE) Nodes.
+func (s *Server) heartbeatEICEInstance(instances *server.EC2Instances) {
+	awsInfo := &types.AWSInfo{
+		AccountID:   instances.AccountID,
+		Region:      instances.Region,
+		Integration: instances.Integration,
+	}
+
+	type existingNodeInfo struct {
+		name       string
+		expiration *time.Time
+	}
+
+	existingEICENodes := make(map[string]existingNodeInfo)
+
+	s.nodeWatcher.GetNodes(s.ctx, func(n services.Node) bool {
+		if !n.IsEICE() {
+			return false
+		}
+		existingEICENodes[n.ServerInfoName()] = existingNodeInfo{
+			name:       n.GetName(),
+			expiration: n.GetMetadata().Expires,
+		}
+		return false
+	})
```

Comment:
> Get respective Server resource from the cache. This resource has expiration info in it.

NodeWatcher only provides iteration methods, there's no method to retrieve a single Server.
Even if it existed, we would use the Server's Name because they get stored in a map whose key is the Server's Name.

The problem is that Server's Name is not deterministic - it is an UUID.
So, when we create a Server object from AWS instance, it will have a random UUID.
To understand if that Server already exists, we have to iterate over all Teleport Servers and see if there's one EICE Node that has the same AccountID/InstanceID (we don't use the Name).
If it exists, we replace the converted Server's name with the found one.

Possible alternatives:
- store another map in NodeWatcher where we use the ServerInfo format as key
- use the ServerInfo format as Node's name

Instead of ServerInfo format it could be any other format that would uniquely identify the node using the AWS AccountID and InstanceID.

I prefer the current version as opposed to the alternatives above.
But please let me know what you think of them and if you have another alternative.

Thank you